dvc/remote/ssh.py

```diff
@@ -184,68 +194,88 @@ def _exec(self, ssh, cmd):
+            except FileNotFoundError:
+                return False
+
     def md5(self, path_info):
-        if path_info["scheme"] != "ssh":
+        if path_info["scheme"] != self.scheme:
             raise NotImplementedError
 
-        ssh = self.ssh(
-            host=path_info["host"],
-            user=path_info["user"],
-            port=path_info["port"],
-        )
-
-        # Use different md5 commands depending on os
-        stdout = self._exec(ssh, "uname").strip()
-        if stdout == "Darwin":
-            md5cmd = "md5"
-            index = -1
-        elif stdout == "Linux":
-            md5cmd = "md5sum"
-            index = 0
-        else:
-            msg = "'{}' is not supported as a remote".format(stdout)
-            raise DvcException(msg)
+        with self.ssh(**path_info) as ssh:
+            # Use different md5 commands depending on the OS:
+            #
+            #  - Darwin's `md5` returns BSD-style checksums by default
+            #  - Linux's `md5sum` needs the `--tag` flag for a similar output
+            #
+            #  Example:
+            #       MD5 (foo.txt) = f3d220a856b52aabbf294351e8a24300
+            #
+            uname = self._exec(ssh, "uname").strip()
+
+            command = {
+                "Darwin": "md5 {path}".format(**path_info),
+                "Linux": "md5sum --tag {path}".format(**path_info),
+            }.get(uname)
+
+            if not command:
+                raise DvcException(
+                    "'{os}' is not supported as a remote".format(os=uname)
+                )
 
-        stdout = self._exec(ssh, "{} {}".format(md5cmd, path_info["path"]))
-        md5 = stdout.split()[index]
-        ssh.close()
+            md5 = self._exec(command).split()[-1]
 
-        assert len(md5) == 32
+            assert len(md5) == 32
 
-        return md5
+            return md5
 
-    def copy(self, from_info, to_info, ssh=None):
-        if from_info["scheme"] != "ssh" or to_info["scheme"] != "ssh":
+    def copy(self, from_info, to_info):
+        if (
+            from_info["scheme"] != self.scheme
+            or to_info["scheme"] != self.scheme
+        ):
             raise NotImplementedError
 
         assert from_info["host"] == to_info["host"]
         assert from_info["user"] == to_info["user"]
+        assert from_info["port"] == to_info["port"]
 
-        s = (
-            ssh
-            if ssh
-            else self.ssh(
-                host=from_info["host"],
-                user=from_info["user"],
-                port=from_info["port"],
-            )
-        )
+        src = from_info["path"]
+        dest = to_info["path"]
+        dest_dir = posixpath.dirname(src)
 
-        dname = posixpath.dirname(to_info["path"])
-        self._exec(s, "mkdir -p {}".format(dname))
-        self._exec(s, "cp {} {}".format(from_info["path"], to_info["path"]))
-
-        if not ssh:
-            s.close()
+        with self.ssh(**from_info) as ssh:
```

Comment:
path_info might contain additional entries that are not related to the ssh connection(e.g. schema, which is one of the reasons why tests fail).